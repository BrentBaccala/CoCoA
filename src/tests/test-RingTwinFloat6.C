//   Copyright (c)  2010  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BigRat.H"
#include "CoCoA/BuildInfo.H"
#include "CoCoA/GlobalManager.H"
#include "CoCoA/RingTwinFloat.H"
#include "CoCoA/VectorOperations.H"

using namespace CoCoA;

#include <iostream>
using std::cout;
using std::cerr;
using std::endl;
#include <vector>
using std::vector;

#define TEST_ASSERT(cond)   do { if (!(cond)) CoCoA::AssertionFailed(#cond, __FILE__, __LINE__); } while(0)

// This test illustrates the "soft transitions" from false to uncertain and from
// uncertain to true in the equality test between twin floats.

void trial(long BitPrec)
{
  const int IterMax = 1000;
  const ring RR = NewRingTwinFloat(BitPrec);

  cout << "RING is " << RR << endl;
  BigRat eps = BigRat(1,1);
  for (long i = 1; i < 99; ++i)
  {
    eps /= 2;

    // Subprogram A
    vector<int> SubprogA(3);
    for (int j=0; j < IterMax; ++j)
    {
      try
      {
        const RingElem X(RR, 1);
        if (X+eps == X)
          ++SubprogA[0];
        else
          ++SubprogA[2];
      }
      catch (const RingTwinFloat::InsufficientPrecision&)
      {
        ++SubprogA[1];
      }
    }

    // Subprogram B
    vector<int> SubprogB(3);
    for (int j=0; j < IterMax; ++j)
    {
      try
      {
        const RingElem X(RR, 1);
        const RingElem Y(RR, 1+eps);
        if (X == Y)
          ++SubprogB[0];
        else
          ++SubprogB[2];
      }
      catch (const RingTwinFloat::InsufficientPrecision&)
      {
        ++SubprogB[1];
      }
    }

    // Subprogram C
    vector<int> SubprogC(3);
    for (int j=0; j < IterMax; ++j)
    {
      try
      {
        const RingElem X(RR, 1);
        const RingElem Y(RR, 1+eps);
        if (IsZero(X-Y))
          ++SubprogC[0];
        else
          ++SubprogC[2];
      }
      catch (const RingTwinFloat::InsufficientPrecision&)
      {
        ++SubprogC[1];
      }
    }


    cout << "precision=" << BitPrec
                   << "  iter=" << i
                   << "  SubprogA=" << SubprogA
                   << "  SubprogB=" << SubprogB
                   << "  SubprogC=" << SubprogC
                   << endl;
  }
  cout << "---------------------------------" << endl;
}


void program()
{
  GlobalManager CoCoAFoundations;

  for (long prec=16; prec <= 32; prec+=16)
    trial(prec);

}


// Use main() to handle any uncaught exceptions and warn the user about them.
int main()
{
  try
  {
    program();
    return 0;
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    cerr << "***ERROR***  UNCAUGHT CoCoA Error";
    ANNOUNCE(cerr, err);
  }
  catch (const std::exception& exc)
  {
    cerr << "***ERROR***  UNCAUGHT std::exception: " << exc.what() << endl;
  }
  catch(...)
  {
    cerr << "***ERROR***  UNCAUGHT UNKNOWN EXCEPTION" << endl;
  }

  BuildInfo::PrintAll(cerr);
  return 1;
}
