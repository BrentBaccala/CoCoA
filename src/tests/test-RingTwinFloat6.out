RING is RingWithID(2,"RingTwinFloat(16)")
precision=16  iter=1  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=2  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=3  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=4  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=5  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=6  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=7  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=8  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=9  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=10  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=11  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=12  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=13  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=14  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=15  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=16  iter=16  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  269,  731]
precision=16  iter=17  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  553,  447]
precision=16  iter=18  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  724,  276]
precision=16  iter=19  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  865,  135]
precision=16  iter=20  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  943,  57]
precision=16  iter=21  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  969,  31]
precision=16  iter=22  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  987,  13]
precision=16  iter=23  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  990,  10]
precision=16  iter=24  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  999,  1]
precision=16  iter=25  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  999,  1]
precision=16  iter=26  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  997,  3]
precision=16  iter=27  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=16  iter=28  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=16  iter=29  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=16  iter=30  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=16  iter=31  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=16  iter=32  SubprogA=[0,  502,  498]  SubprogB=[0,  478,  522]  SubprogC=[0,  1000,  0]
precision=16  iter=33  SubprogA=[0,  732,  268]  SubprogB=[0,  852,  148]  SubprogC=[0,  1000,  0]
precision=16  iter=34  SubprogA=[0,  879,  121]  SubprogB=[0,  975,  25]  SubprogC=[0,  1000,  0]
precision=16  iter=35  SubprogA=[0,  939,  61]  SubprogB=[0,  993,  7]  SubprogC=[0,  1000,  0]
precision=16  iter=36  SubprogA=[0,  970,  30]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=16  iter=37  SubprogA=[0,  982,  18]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=16  iter=38  SubprogA=[0,  992,  8]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=16  iter=39  SubprogA=[0,  999,  1]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=16  iter=40  SubprogA=[0,  996,  4]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=16  iter=41  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=16  iter=42  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=16  iter=43  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=16  iter=44  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=16  iter=45  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=16  iter=46  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=16  iter=47  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=16  iter=48  SubprogA=[486,  514,  0]  SubprogB=[511,  489,  0]  SubprogC=[497,  503,  0]
precision=16  iter=49  SubprogA=[733,  267,  0]  SubprogB=[869,  131,  0]  SubprogC=[867,  133,  0]
precision=16  iter=50  SubprogA=[873,  127,  0]  SubprogB=[978,  22,  0]  SubprogC=[967,  33,  0]
precision=16  iter=51  SubprogA=[947,  53,  0]  SubprogB=[994,  6,  0]  SubprogC=[994,  6,  0]
precision=16  iter=52  SubprogA=[964,  36,  0]  SubprogB=[998,  2,  0]  SubprogC=[998,  2,  0]
precision=16  iter=53  SubprogA=[976,  24,  0]  SubprogB=[998,  2,  0]  SubprogC=[999,  1,  0]
precision=16  iter=54  SubprogA=[991,  9,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=55  SubprogA=[999,  1,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=56  SubprogA=[997,  3,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=57  SubprogA=[999,  1,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=58  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=59  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=60  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=61  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=62  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=63  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=64  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=65  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=66  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=67  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=68  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=69  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=70  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=71  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=72  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=73  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=74  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=75  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=76  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=77  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=78  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=79  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=80  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=81  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=82  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=83  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=84  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=85  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=86  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=87  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=88  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=89  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=90  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=91  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=92  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=93  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=94  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=95  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=96  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=97  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=16  iter=98  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
---------------------------------
RING is RingWithID(3,"RingTwinFloat(32)")
precision=32  iter=1  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=2  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=3  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=4  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=5  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=6  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=7  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=8  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=9  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=10  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=11  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=12  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=13  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=14  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=15  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=16  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=17  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=18  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=19  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=20  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=21  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=22  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=23  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=24  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=25  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=26  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=27  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=28  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=29  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=30  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=31  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  0,  1000]
precision=32  iter=32  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  230,  770]
precision=32  iter=33  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  562,  438]
precision=32  iter=34  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  769,  231]
precision=32  iter=35  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  859,  141]
precision=32  iter=36  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  946,  54]
precision=32  iter=37  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  963,  37]
precision=32  iter=38  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  988,  12]
precision=32  iter=39  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  992,  8]
precision=32  iter=40  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  995,  5]
precision=32  iter=41  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=42  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=43  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=44  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  999,  1]
precision=32  iter=45  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=46  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=47  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=48  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=49  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=50  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=51  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=52  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=53  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=54  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=55  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=56  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=57  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=58  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=59  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=60  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=61  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=62  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=63  SubprogA=[0,  0,  1000]  SubprogB=[0,  0,  1000]  SubprogC=[0,  1000,  0]
precision=32  iter=64  SubprogA=[0,  490,  510]  SubprogB=[0,  478,  522]  SubprogC=[0,  1000,  0]
precision=32  iter=65  SubprogA=[0,  732,  268]  SubprogB=[0,  873,  127]  SubprogC=[0,  1000,  0]
precision=32  iter=66  SubprogA=[0,  885,  115]  SubprogB=[0,  972,  28]  SubprogC=[0,  1000,  0]
precision=32  iter=67  SubprogA=[0,  935,  65]  SubprogB=[0,  994,  6]  SubprogC=[0,  1000,  0]
precision=32  iter=68  SubprogA=[0,  972,  28]  SubprogB=[0,  999,  1]  SubprogC=[0,  1000,  0]
precision=32  iter=69  SubprogA=[0,  980,  20]  SubprogB=[0,  999,  1]  SubprogC=[0,  1000,  0]
precision=32  iter=70  SubprogA=[0,  996,  4]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=32  iter=71  SubprogA=[0,  997,  3]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=32  iter=72  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=32  iter=73  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=32  iter=74  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=32  iter=75  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=32  iter=76  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=32  iter=77  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=32  iter=78  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=32  iter=79  SubprogA=[0,  1000,  0]  SubprogB=[0,  1000,  0]  SubprogC=[0,  1000,  0]
precision=32  iter=80  SubprogA=[501,  499,  0]  SubprogB=[490,  510,  0]  SubprogC=[473,  527,  0]
precision=32  iter=81  SubprogA=[751,  249,  0]  SubprogB=[882,  118,  0]  SubprogC=[886,  114,  0]
precision=32  iter=82  SubprogA=[879,  121,  0]  SubprogB=[968,  32,  0]  SubprogC=[966,  34,  0]
precision=32  iter=83  SubprogA=[941,  59,  0]  SubprogB=[990,  10,  0]  SubprogC=[994,  6,  0]
precision=32  iter=84  SubprogA=[967,  33,  0]  SubprogB=[998,  2,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=85  SubprogA=[983,  17,  0]  SubprogB=[999,  1,  0]  SubprogC=[999,  1,  0]
precision=32  iter=86  SubprogA=[995,  5,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=87  SubprogA=[994,  6,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=88  SubprogA=[998,  2,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=89  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=90  SubprogA=[999,  1,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=91  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=92  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=93  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=94  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=95  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=96  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=97  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
precision=32  iter=98  SubprogA=[1000,  0,  0]  SubprogB=[1000,  0,  0]  SubprogC=[1000,  0,  0]
---------------------------------
